# Empirical expansion review

## tl;dr
The September 2026 snapshot expands lobbying history to 2003-2008 and acquires 229 reports for four candidate PACs. A reviewed registration alias restores fifteen AdvaMed filings. Separate filing-grain metadata and two original-form reviews reveal incompatible accounting methods and potential income/expense double counting. Of 48 expected committee-half-years, 36 have usable outcomes, including one source-reviewed supplemental-filing decision; twelve remain unresolved. These are source and design inputs, not identified substitution effects. The archived procurement audit finds 243,700 blank award types and 4,505,798 missing offer counts among 6,449,101 rows. A competition-after-exclusion flag is not SAM vendor status. Comment attribution and procurement source/linkage gaps remain unresolved.

A separate heavy-duty Phase 3 pilot records two agency-reproduced requests with explicit named responses: one clarification and one rejected error claim. Original attachments and independent review remain unresolved. These selected cases establish documentary responses, not an uptake rate or causal effect. A retrospective review of the complete six-page section now inventories seven organization blocks and eleven distinct requests, including those same two requests. The inventory distinguishes three named responses, four links through the named summary and collective response, one section-resolution-only comparison, and three requests without a separate disposition in this section. It remains agency-selected evidence, not a docket denominator.

The selected GAO decision's thirteen listed docket entries now have source-reviewed filing dates: six June 24, four June 26 and three August 2, 2024. They remain one consolidated decision. Multiple solicitation/service-area scopes and missing original-action mapping fields prevent award-specific timing promotion.

A targeted cross-section follow-up now links DTNA's existing averaging-set request to a named summary and collective response: partial alignment with interim flexibility, not full acceptance. The proposal already solicited the option. Four facets remain one request, and the eleven-request baseline is unchanged.

All three flagged LDA posting-date anomalies now have full embedded-cover reviews and matching registration components. Source receipt stamps postdate the reporting periods but do not replace the earlier API posting dates. Two covers add Method A observations; one API zero is a below-threshold income disclosure, not an observed point zero. Raw metadata, version-ordering restrictions and control-assignment boundaries are preserved.

The LDA family queue covers 22 multi-record candidate groups containing 46 filings, with eight different-amount flags, three posting ties and two latest-posting missing-amount flags. A selected APGA amendment contains only one distinct issue image, not a financial cover. Both energy-page scans name FERC but their API contact lists do not reproduce that agency list. That APGA review selects no final version, spending amount or exposure classification.

A follow-up reviews all 21 frozen NVG filings: seventeen full embedded PDF covers and four electronic HTML identity/timing sections. Six name America Votes, including one registration, and are ineligible for AAJ attribution. Five mixed-client candidate groups are not mergeable. Two cover years disagree with API years; Senate suffixes cannot be globally repaired. The earlier America Votes House-ID transcription is corrected to 35960015 against the unchanged source image. The other fifteen records are not cleared as final actor-period outcomes. Raw metadata remains unchanged, and independent review is pending.

The original-action follow-up adds four contract downloads and 26 history rows. Two explicit service-area labels and two rounded-value candidates support sixteen provisional pairs across thirteen dockets. The candidate VISN 8 row reports four offers versus five timely proposals in the decision; the discrepancy remains unresolved. No baseline timing or causal claims are promoted.

A separate AAJ 2005H1 pair supports one reported $4,020,000 Method A expense through matching covers and an explicit unchanged-expense letter. All fourteen distinct downloaded images were reviewed. Empty amendment API activities and two original contact discrepancies remain unresolved. This field-specific adjudication does not select a whole final filing or clear actor totals, controls or independent review.

The Benefits Council paper review matches all 24 frozen reports to a separate 280-record filing inventory and recovers 21 current plus three superseded version flags. Seven selected report PDFs reveal three native-form discrepancies and three downloaded packets without a disbursement-summary page. Version recovery adds no complete half-year: eleven periods fail on missing outcomes and 2006H1 first fails on its source-confirmed date gap. Raw dates/amounts and the 36 usable half-years remain unchanged; independent review remains pending.

## Context & Methods
This offline companion reruns the repository's evidence audit and checks the prepared PAC series against its source report periods. It also distinguishes two public-letter positions from a verified docket copy or observed agency uptake. It makes no network requests and requires no credentials. Run from this notebook's directory or the repository root using Python 3.

### Key Assumptions
Exact-name LDA links, one registration-specific reviewed alias and historical PAC affiliations do not independently validate the actor spine. Native reporting periods are not independent quarters. Repeated issue rows are not separate monetary observations. A missing amount is not zero. One API-reported zero is reviewed as censored income; other zeros still require original-form adjudication. Organizational expenses may include outside-firm income reports, and reporting methods may cover different activity scopes. Event labels use September 14, 2007 as an enactment convention, not a universal treatment date: House travel restrictions began earlier, so 2007H1 is not an untreated period for that contrast. Read `docs/substitution-study-redesign.md`, `docs/comment-uptake-pilot.md`, and `docs/procurement-source-reconciliation.md` for source URLs and remaining requirements.

## Data
Inputs are the public CSV and JSON snapshots under `data/calibration/first-wave/`. The audit profiles LDA/FEC coverage and adjudications, the comment pilot, and both the GAO discovery worklist and separate four-award linkage pilot. LDA filing metadata, alias reviews and original-form reviews preserve separate source dollars, accounting methods, registration IDs and review scope. The GAO pilot preserves public search responses, projected award details and four archived bulk rows, with source URLs and hashes. Its saved extract is checked against the committed bulk manifest; this notebook does not require or independently re-scan the ignored 6.4-million-row bulk file. Input paths remain repository-relative.

In [1]:
import csv
import importlib.util
import json
from pathlib import Path

root = Path.cwd()
if root.name == 'notebooks':
    root = root.parent
assert (root / 'scripts/audit-empirical-expansion.py').is_file(), 'Run inside this repository'

def load_script(name, filename):
    spec = importlib.util.spec_from_file_location(name, root / 'scripts' / filename)
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module

audit = load_script('expansion_audit', 'audit-empirical-expansion.py')
periods = load_script('fec_periods', 'prepare-substitution-fec-periods.py')
findings = audit.audit()

## Results
### Reproduce the prepared series
Source-marked latest, non-amended reports enter complete half-years, with one explicitly recorded supplemental-loan-paperwork adjudication. The preparation retains every expected half-year and quarantines missing/nonfinite outcomes, gaps, overlaps, straddling reports and unresolved alternatives. The reform-straddling half-year is retained but excluded from event contrasts. The strict source-flag-only sensitivity omits the reviewed exception.

In [2]:
raw_reports = audit.read('substitution-fec-report-panel.csv')
cohort = audit.read('substitution-fec-acquisition-cohort.csv')
adjudications = audit.read('substitution-fec-version-adjudications.csv')
paper_review = json.loads((audit.DATA / 'substitution-fec-paper-review.json').read_text())
prepared, coverage = periods.prepare_with_coverage(raw_reports, cohort, adjudications, paper_review)
strict_prepared, _ = periods.prepare_with_coverage(raw_reports, cohort)
assert prepared == audit.read('substitution-fec-halfyear-panel.csv')
assert coverage == audit.read('substitution-fec-halfyear-coverage.csv')
print(f'{len(raw_reports)} raw reports; {len(prepared)} of {len(coverage)} complete half-years; {len(strict_prepared)} without manual adjudication')
for finding in findings:
    print(f"{finding['item']}: {finding['status']}")
    print(finding['evidence'])

229 raw reports; 36 of 48 complete half-years; 35 without manual adjudication
expanded-lda-history: api_attribution_identity_unresolved
issueRows=1119; filingUUIDs=427; actors=6; years=2003,2004,2005,2006,2007,2008; observedPrePeriodsByActor=cand-372cc95f9387:9;cand-5d1da86118e5:9;cand-9948f2974958:9;cand-b905d6833296:9;cand-d5522e62fad7:9;cand-f7178708cc78:9
lda-posting-date-anomaly: source_reviewed_dates_not_orderable
postingBeforeCoveredPeriod=3; reviewedEmbeddedCovers=3; receiptDatesAfterPeriod=3; registrationComponentMatches=3; differentRegistrantNames=2; reviewedExpenseMethods={'A': 2}; censoredIncomeDisclosures=1; sourceMissingTerminationDates=1; correctedApiPostingDates=0; amendmentOrderingCleared=False; controlAssignmentCleared=False; causalEffect=not_identified
lda-measurement-comparability: source_measures_not_comparable_totals
filings=427; amountKinds={'income': 267, 'neither': 71, 'expenses': 89}; expenseMethodMissing=63; sourceZeroFilings=25; incomeExpenseOverlapActorPeri

### Reconcile Benefits Council paper versions without filling financial gaps

`substitution-fec-paper-review.json` preserves two endpoint projections, the complete acquired filing-image inventory, and seven selected report-PDF reviews plus one FEC letter. The join uses beginning-image number and verifies committee, file number, report type, coverage/receipt dates and PDF URL. It recovers only `mostRecent`; neither amounts nor dates are corrected. The reports endpoint's `isAmended` remains separate from the unavailable filings projection.

The original late-2008 forms are Form 3 despite the API's F3X classification. Their amendments use Form 3X following the May 15, 2009 letter. Three downloaded packets lack disbursement-summary pages; this does not establish which pages were originally filed. Form, printed line, column and actual PDF page must be distinguished. Blank cells and year-to-date amounts are not numeric period zeros.

This cell reproduces the version-only decision and compares it with the prior coverage. The following optional checks validate available acquisition projections and PDF bytes/page counts. Missing files or Poppler are reported as checks not run. Hashes authenticate saved bytes, not manual interpretations or historical completeness. Primary links and exact page scopes are in the ledger and `docs/substitution-study-redesign.md`.

In [3]:
paper_module = load_script('fec_paper', 'review-substitution-fec-paper.py')
paper_inputs_before = json.dumps([raw_reports, paper_review], sort_keys=True)
version_flags, paper_checks = paper_module.validate_review(raw_reports, paper_review)
assert len(version_flags) == 24 and sum(version_flags.values()) == 21
assert paper_checks['filingInventoryRows'] == 280
assert paper_checks['nativeFormTypeDiscrepancies'] == 3
assert paper_checks['packetsWithoutDisbursementSummary'] == 3
prior_prepared, prior_coverage = periods.prepare_with_coverage(raw_reports, cohort, adjudications)
assert prior_prepared == prepared and len(prepared) == 36 and len(coverage) == 48
benefits_coverage = [row for row in coverage if row['committeeId'] == paper_review['committeeId']]
assert len(benefits_coverage) == 12
assert sum(row['reasonCodes'] == 'missing_outcome' for row in benefits_coverage) == 11
assert [row['halfYear'] for row in benefits_coverage if row['reasonCodes'] == 'gap_or_overlap'] == ['2006H1']
assert all(row['status'] == 'unresolved' and not row['unknownVersionRecordIds'] for row in benefits_coverage)
assert sum(int(row['eligibleReportCount']) for row in benefits_coverage) == 21
assert paper_inputs_before == json.dumps([raw_reports, paper_review], sort_keys=True)
print(paper_checks)
for row in benefits_coverage:
    print(row['halfYear'], row['reasonCodes'], 'eligible versions:', row['eligibleReportCount'])
print('Version-metadata obstacle resolved for frozen frame; zero additional complete financial half-years.')

{'reportRowsMatched': 24, 'filingInventoryRows': 280, 'filingsLatestTrue': 21, 'filingsLatestFalse': 3, 'reportEndpointUnknownLatest': 24, 'reviewedReportPdfs': 7, 'nativeFormTypeDiscrepancies': 3, 'packetsWithoutDisbursementSummary': 3, 'reviewedBlankAmountFields': 5, 'rawAmountsPromoted': 0, 'datesCorrected': 0, 'independentReviewStatus': 'pending', 'causalEffect': 'not_identified'}
2003H1 missing_outcome eligible versions: 1
2003H2 missing_outcome eligible versions: 1
2004H1 missing_outcome eligible versions: 2
2004H2 missing_outcome eligible versions: 3
2005H1 missing_outcome eligible versions: 1
2005H2 missing_outcome eligible versions: 1
2006H1 gap_or_overlap eligible versions: 2
2006H2 missing_outcome eligible versions: 3
2007H1 missing_outcome eligible versions: 1
2007H2 missing_outcome eligible versions: 1
2008H1 missing_outcome eligible versions: 2
2008H2 missing_outcome eligible versions: 3
Version-metadata obstacle resolved for frozen frame; zero additional complete financi

In [4]:
import hashlib
import shutil
import subprocess

def verify_fec_paper_files(review, raw_directory):
    verified, unavailable = [], []
    for filename, acquisition in [('report-projections.json', review['reportAcquisition']),
                                  ('filing-projections.json', review['filingsQuery'])]:
        source = raw_directory / filename
        if not source.is_file():
            unavailable.append(filename + ': acquisition projection unavailable')
            continue
        content = source.read_bytes()
        assert hashlib.sha256(content).hexdigest() == acquisition['projectionFileSha256']
        payload = json.loads(content)
        assert payload['retrievedAt'] == acquisition['retrievedAt']
        if filename == 'report-projections.json':
            assert sorted(payload['records'], key=lambda row: row['sourceRecordId']) == review['reportProjections']
            assert all(query['pagination'] == 'all_pages_via_fetch_all' for query in payload['queries'])
        else:
            inventory = [row['beginning_image_number'] for row in payload['records']]
            assert len(inventory) == len(set(inventory)) == payload['count'] == review['filingsQuery']['count']
            assert sorted(inventory) == review['filingsQuery']['beginningImageInventory']
            assert payload['pagination'] == 'all_pages_via_fetch_all'
            by_image = {row['beginning_image_number']: row for row in payload['records']}
            for projection in review['matchedFilingProjections']:
                assert all(by_image[projection['beginning_image_number']][key] == value
                           for key, value in projection.items())
        verified.append(filename + ': projection hash and saved contents')
    documents = [(document['label'] + '.pdf', document) for document in review['documents']]
    documents.append(('rfai-2008q3.pdf', review['formCorrectionLetter']))
    for filename, document in documents:
        source = raw_directory / filename
        if not source.is_file():
            unavailable.append(filename + ': PDF unavailable')
            continue
        assert hashlib.sha256(source.read_bytes()).hexdigest() == document['pdfSha256']
        verified.append(filename + ': PDF hash')
        if not shutil.which('pdfinfo'):
            unavailable.append(filename + ': page count requires Poppler pdfinfo')
            continue
        info = subprocess.run(['pdfinfo', str(source)], check=True, capture_output=True, text=True).stdout
        pages = next(line for line in info.splitlines() if line.startswith('Pages:'))
        assert int(pages.split(':', 1)[1]) == document['pdfPages']
        verified.append(filename + ': PDF page count')
    return verified, unavailable

paper_verified, paper_unavailable = verify_fec_paper_files(
    paper_review, root / 'data/raw/substitution-fec-paper-review-2026-09-13')
for detail in paper_verified:
    print('Verified:', detail)
for detail in paper_unavailable:
    print('Not run:', detail)
print('These are metadata projections, not raw HTTP bodies. Independent interpretation remains pending.')

Verified: report-projections.json: projection hash and saved contents
Verified: filing-projections.json: projection hash and saved contents
Verified: 2006q1.pdf: PDF hash
Verified: 2006q1.pdf: PDF page count
Verified: 2008post-amendment.pdf: PDF hash
Verified: 2008post-amendment.pdf: PDF page count
Verified: 2008post-original.pdf: PDF hash
Verified: 2008post-original.pdf: PDF page count
Verified: 2008q3-amendment.pdf: PDF hash
Verified: 2008q3-amendment.pdf: PDF page count
Verified: 2008q3-original.pdf: PDF hash
Verified: 2008q3-original.pdf: PDF page count
Verified: 2008ye-amendment.pdf: PDF hash
Verified: 2008ye-amendment.pdf: PDF page count
Verified: 2008ye-original.pdf: PDF hash
Verified: 2008ye-original.pdf: PDF page count
Verified: rfai-2008q3.pdf: PDF hash
Verified: rfai-2008q3.pdf: PDF page count
These are metadata projections, not raw HTTP bodies. Independent interpretation remains pending.


### Inspect source and reviewed accounting methods separately
One metadata row represents one filing, not one issue or one organization-period. The audit checks exact joins, amount conversion, projected-source fingerprints and alias scope. Original-form method observations remain separate from raw API fields and are not carried across periods. The two reviewed covers are not a representative sample or an amendment-family audit.

In [5]:
metadata = audit.read('substitution-lda-filing-metadata.csv')
filing_reviews = audit.read('substitution-lda-filing-reviews.csv')
metadata_by_uuid = {row['filingUuid']: row for row in metadata}
assert len(metadata_by_uuid) == len(metadata)
for review in filing_reviews:
    source = metadata_by_uuid[review['filingUuid']]
    assert review['sourceFingerprint'] == source['sourceFingerprint']
    method = source['expensesMethod'] or 'missing'
    print(f"{source['primaryName']}: {source['filingYear']} {source['filingPeriod']}; API method {method}; original form method {review['formExpensesMethod']}; reviewed page {review['pagesReviewed']}")

ADVANCED MEDICAL TECHNOLOGY ASSOCIATION: 2003 mid_year; API method missing; original form method C; reviewed page 1
AMERICAN ASSOCIATION FOR JUSTICE: 2003 mid_year; API method missing; original form method A; reviewed page 1


### Reconcile every flagged LDA posting date against the full cover
The anomaly-selected `substitution-lda-date-reviews.json` covers all three unique filings with an API posting date before the covered period. Full embedded cover scans restore fields clipped by the ordinary PDF rendering. The source record preserves public API projections, PDF and image hashes, scan identifiers, field-level observations, and pending independent review. It does not audit the other 424 filing dates or every page/version of these three filings.

Receipt stamps, posting timestamps and termination dates are different fields. No replacement timestamp or timezone is invented. Matching the Senate registrant and client-relationship components links each source registration, not an exhaustive actor identity or valid control. Two covers add filing-specific Method A observations; one firm's below-$10,000 income disclosure is censored, not a point zero. This offline check validates the saved projections and coding boundaries, not independent source interpretation.

In [6]:
date_sources = json.loads((audit.DATA / 'substitution-lda-date-reviews.json').read_text())
metadata_before = json.dumps(metadata, sort_keys=True)
date_checks = audit.lda_date_review_diagnostics(metadata, date_sources)
assert date_checks['postingBeforeCoveredPeriod'] == date_checks['reviewedEmbeddedCovers'] == 3
assert date_checks['registrationComponentMatches'] == 3
assert date_checks['reviewedExpenseMethods'] == {'A': 2}
assert date_checks['censoredIncomeDisclosures'] == date_checks['sourceMissingTerminationDates'] == 1
assert date_checks['correctedApiPostingDates'] == 0
assert not date_checks['amendmentOrderingCleared'] and not date_checks['controlAssignmentCleared']
assert json.dumps(metadata, sort_keys=True) == metadata_before
for review in date_sources['reviews']:
    source = metadata_by_uuid[review['filingUuid']]
    form = review['form']
    print(f"{form['senateId']}: {form['year']} {form['period']}; API posted {source['dtPosted'][:10]}; cover receipt {form['receiptDate']}")
    print(f"  {form['amountKind']}: {form['amountDisclosure']}; API method {source['expensesMethod'] or 'missing'}; cover method {form['expensesMethod'] or 'not applicable'}")
    if form['amountDisclosure'] == 'less_than_threshold_censored':
        assert form['reportedAmountDollars'] is None and source['incomeDollars'] == '0.00'
        print(f"  No point amount; reported income below ${form['incomeUpperBoundExclusiveDollars']}; scanned termination {form['terminationDate']}; API termination missing.")
print('Three reviewed covers; raw metadata unchanged; no corrected posting times, amendment order or controls promoted.')

74077-12: 2003 mid_year; API posted 2002-08-29; cover receipt 2003-07-31
  expenses: numeric_amount_threshold_boxes_unchecked; API method missing; cover method A
4733-12: 2003 year_end; API posted 2003-02-24; cover receipt 2004-02-24
  expenses: at_least_threshold_rounded; API method missing; cover method A
35024-90: 2004 year_end; API posted 1995-05-19; cover receipt 2005-01-19
  income: less_than_threshold_censored; API method missing; cover method not applicable
  No point amount; reported income below $10000.00; scanned termination 2005-01-05; API termination missing.
Three reviewed covers; raw metadata unchanged; no corrected posting times, amendment order or controls promoted.


### Inspect candidate filing families without choosing final versions
`substitution-lda-family-queue.csv` covers every multi-record group in the frozen metadata, not all historical versions. Registrations are excluded; singleton reports are not certified complete. Grouping uses registration components and native periods, with actor/client consistency checked. Source amounts remain separate income or expense disclosures. Flags overlap and are review leads, not verified errors. A No Activity label does not imply zero income.

The source JSON retains one terminal APGA query, native API activity projections, five distinct embedded-image hashes and page mappings. The API is not a faithful original/amended field history in this case. Scans show a blank-to-handwritten lobbyist field and FERC contacts, whereas the API already names the lobbyist in both versions and supplies different contact lists. The original financial cover remains linked to its separate date review; missing amendment amounts are neither zero nor automatically inherited. Direct source URLs, historical-amendment limitations and independent-review requirements are in `docs/substitution-study-redesign.md`. These offline checks verify saved consistency, not historical packet completeness or independent source interpretation.

The initial NVG cover pair came from an amount-conflict/posting-tie group, with a separate fourteen-record terminal query verifying only those two metadata projections. The follow-up covers all 21 NVG records in the frozen metadata acquisition, not the remaining twelve query results or all public NVG filings. The source ledger binds seventeen complete embedded cover images and four HTML field groups to their original fingerprints and document hashes. Six America Votes records cannot be attributed to AAJ; the remaining names do not independently establish historical links. Review scope excludes other pages, financial-version adjudication and causal/control assignment. The House-ID transcription correction is preserved in the review history.

In [7]:
family_spec = importlib.util.spec_from_file_location('lda_family_review', root / 'scripts/review-substitution-lda-families.py')
family_module = importlib.util.module_from_spec(family_spec)
family_spec.loader.exec_module(family_module)
family_sources = json.loads((audit.DATA / 'substitution-lda-family-review.json').read_text())
metadata_before_family = json.dumps(metadata, sort_keys=True)
family_queue, family_checks = family_module.validate_review(metadata, family_sources, date_sources)
assert family_queue == audit.read('substitution-lda-family-queue.csv')
assert family_checks['candidateGroups'] == 364
assert family_checks['sourceReviewedGroups'] == 7
assert family_checks['multiRecordGroups'] == 22 and family_checks['multiRecordFilings'] == 46
assert family_checks['groupsWithDifferentSourceAmounts'] == 8
assert family_checks['groupsWithPostingTies'] == 3
assert family_checks['groupsWithLatestMissingEarlierAmount'] == 2
assert family_checks['groupsWithNoActivityNonzero'] == 1
assert family_checks['singletonAmendmentGroups'] == 1
assert family_checks['reviewedDistinctImages'] == 36 and family_checks['energyAgencyMismatches'] == 2
assert family_checks['initialIdentityReviewedCovers'] == 2
assert family_checks['initialDifferentClientCovers'] == 1
assert family_checks['initialRegistrationSuffixDiscrepancies'] == 1
assert family_checks['initialUnlabelledAmendedCovers'] == 1
conflicted = [row for row in family_queue if row['reviewStatus'] == 'source_identity_conflict_not_mergeable']
assert len(conflicted) == 5 and all(r['registrantApiId'] == '76833' for r in conflicted)
assert family_checks['nvgHistoryReviewedFilings'] == 21
assert family_checks['nvgHistoryPdfCovers'] == 17 and family_checks['nvgHistoryHtmlForms'] == 4
assert family_checks['nvgDifferentClientFilings'] == 6 and family_checks['nvgDifferentClientActivityFilings'] == 5
assert family_checks['nvgSourceYearDisagreements'] == 2
assert family_checks['nvgSourceSenateIdDisagreements'] == 6 and family_checks['nvgSourceSenateIdBlank'] == 5
assert family_checks['finalVersionsSelected'] == 0 and family_checks['causalEffect'] == 'not_identified'
assert json.dumps(metadata, sort_keys=True) == metadata_before_family
for key, value in family_checks.items():
    print(f'{key}: {value}')
print('Selected APGA candidate group:')
for row in family_queue:
    if (row['registrantApiId'], row['clientRelationshipId'], row['filingYear'], row['filingPeriod']) == ('74077', '12', '2003', 'mid_year'):
        print({key: row[key] for key in ('filings', 'filingTypes', 'latestPostingMissingWithEarlierAmount', 'reviewStatus')})
print('APGA original disclosure remains at original-filing grain; amendment amount and final economic observation are unassigned.')
print('All 21 frozen NVG records reviewed at identity/timing scope; six America Votes filings are ineligible for AAJ attribution, and five candidate groups are not mergeable. No raw row was remapped.')
for review in family_sources['nvgHistoryReview']['records']:
    form = review['form']
    print(review['filingUuid'][:8], form['clientName'], form['senateId'], review['identityDisposition'])

sourceFilings: 427
registrationRecordsExcluded: 39
candidateGroups: 364
multiRecordGroups: 22
singletonAmendmentGroups: 1
multiRecordFilings: 46
groupsWithoutAmendmentLabel: 10
groupsWithDifferentSourceAmounts: 8
groupsWithPostingTies: 3
groupsWithLatestMissingEarlierAmount: 2
groupsWithNoActivityNonzero: 1
sourceReviewedGroups: 7
apgaQueriedApiFilings: 2
nvgQueryReturnedFilings: 14
identityApiFilingsChecked: 2
apgaReviewedDistinctImages: 5
reviewedDistinctImages: 36
energyAgencyMismatches: 2
initialIdentityReviewedCovers: 2
initialDifferentClientCovers: 1
initialRegistrationSuffixDiscrepancies: 1
initialUnlabelledAmendedCovers: 1
nvgHistoryReviewedFilings: 21
nvgHistoryPdfCovers: 17
nvgHistoryHtmlForms: 4
nvgDifferentClientFilings: 6
nvgDifferentClientActivityFilings: 5
nvgMixedClientCandidateGroups: 5
nvgSourceYearDisagreements: 2
nvgSourceSenateIdDisagreements: 6
nvgSourceSenateIdBlank: 5
expenseFamiliesResolved: 1
aajReviewedDistinctImages: 14
aajOriginalApiActivityRows: 7
aajAmend

### Resolve one reported expense field without selecting a whole filing
The preselected AAJ 2005H1 pair has matching self-client covers, Method A and reported expenses. Its letter explicitly says expenses are unchanged after adding an omitted TOR item. The source ledger records original printed page 10 (PDF pages 19-20), amended printed page 10 (PDF pages 3-4), and the letter (PDF pages 5-6). This supports one reported expense observation, not adding versions or independently audited spending. Source receipt dates, the letter date and API posting timestamps remain distinct.

The original API has seven activity rows while the amendment API has none despite its TOR addendum. The original TAX and INS source pages name only the House; their API rows also name the Senate. Preserve these discrepancies independently of the expense decision. Source URLs and all image/page references are in `substitution-lda-family-review.json` under `aajExpenseReview` and in `docs/substitution-study-redesign.md`.

The first cell reproduces the committed expense row after the full family validation above. The optional raw check authenticates available local query/PDF bytes, PDF image-object/page coverage, and all fourteen distinct image pairs. Missing raw downloads or Poppler are explicitly reported as checks not run, not successes. No network access or independent source interpretation is implied.

In [8]:
expense_review = family_sources['aajExpenseReview']
expense_rows, expense_checks = family_module.expense_adjudications(metadata, expense_review)
assert expense_rows == audit.read('substitution-lda-expense-adjudications.csv')
assert len(expense_rows) == 1
assert expense_rows[0]['expensesDollars'] == '4020000.00'
assert expense_rows[0]['expensesMethod'] == 'A'
assert expense_rows[0]['actorPeriodAggregationCleared'] == 'false'
assert expense_checks['aajReviewedDistinctImages'] == 14
assert expense_checks['aajOriginalApiActivityRows'] == 7
assert expense_checks['aajAmendmentApiActivityRows'] == 0
assert expense_checks['aajOriginalContactDiscrepancies'] == 2
assert expense_checks['independentlyReviewedExpenseFamilies'] == 0
assert json.dumps(metadata, sort_keys=True) == metadata_before_family
print({key: expense_rows[0][key] for key in ('adjudicationId', 'expensesDollars', 'expensesMethod',
                                          'reviewStatus', 'independentReviewStatus', 'actorPeriodAggregationCleared')})
print('One reported expense observation; no version sum, whole-filing replacement or control assignment.')
print('Original/amendment API activity rows: 7/0; original source-contact discrepancies: 2.')

{'adjudicationId': 'aaj-2005h1-expense', 'expensesDollars': '4020000.00', 'expensesMethod': 'A', 'reviewStatus': 'source_supported_expense_only', 'independentReviewStatus': 'pending', 'actorPeriodAggregationCleared': 'false'}
One reported expense observation; no version sum, whole-filing replacement or control assignment.
Original/amendment API activity rows: 7/0; original source-contact discrepancies: 2.


In [9]:
import hashlib
import shutil
import struct
import subprocess
import tempfile

def verify_expense_raw(review, raw_directory):
    verified, unavailable = [], []
    query_path = raw_directory / 'aaj-query.json'
    if query_path.is_file():
        payload = query_path.read_bytes()
        assert hashlib.sha256(payload).hexdigest() == review['query']['responseSha256']
        query = json.loads(payload)
        assert all(query[field] == review['query'][field] for field in ('count', 'next', 'previous'))
        assert [family_module.api_projection(row) for row in query['results']] == review['query']['records']
        verified.append('query hash, pagination and both complete saved projections')
    else:
        unavailable.append('query raw bytes absent')
    for document in review['documents']:
        role = document['role']
        pdf_path = raw_directory / f'aaj-{role}.pdf'
        if not pdf_path.is_file():
            unavailable.append(f'{role} PDF raw bytes absent')
            continue
        assert hashlib.sha256(pdf_path.read_bytes()).hexdigest() == document['pdfSha256']
        verified.append(f'{role} PDF hash')
        if not shutil.which('pdfimages') or not shutil.which('pdfinfo'):
            unavailable.append(f'{role} image/page check requires Poppler pdfimages and pdfinfo')
            continue
        info = subprocess.run(['pdfinfo', str(pdf_path)], check=True, capture_output=True, text=True).stdout
        page_line = next(line for line in info.splitlines() if line.startswith('Pages:'))
        assert int(page_line.split(':', 1)[1]) == document['pdfPages']
        listing = subprocess.run(['pdfimages', '-list', str(pdf_path)], check=True,
                                 capture_output=True, text=True).stdout
        image_rows = [line.split() for line in listing.splitlines()
                      if line.split() and line.split()[0].isdigit()]
        assert len(image_rows) == 2 * len(document['images']) == document['pdfPages']
        with tempfile.TemporaryDirectory(prefix='lda-expense-images-') as temporary:
            prefix = Path(temporary) / role
            subprocess.run(['pdfimages', '-png', str(pdf_path), str(prefix)], check=True, capture_output=True)
            assert len(list(Path(temporary).iterdir())) == len(image_rows)
            for index, source_image in enumerate(document['images']):
                for duplicate, page in enumerate(source_image['pdfPages']):
                    sequence = 2 * index + duplicate
                    record = image_rows[sequence]
                    assert (int(record[0]), int(record[1]), record[2]) == (page, sequence, 'image')
                    assert [int(record[3]), int(record[4])] == [source_image['width'], source_image['height']]
                    assert ' '.join(record[10:12]) == source_image['objectId']
                    png = Path(f'{prefix}-{sequence:03}.png').read_bytes()
                    assert png[:8] == b'\x89PNG\r\n\x1a\n'
                    assert struct.unpack('>II', png[16:24]) == (source_image['width'], source_image['height'])
                    assert hashlib.sha256(png).hexdigest() == source_image['pngSha256']
            verified.append(f"{role}: {len(document['images'])} distinct image pairs, dimensions, objects and complete PDF pages")
    return verified, unavailable

expense_raw_verified, expense_raw_unavailable = verify_expense_raw(
    expense_review, root / 'data/raw/aaj-2005h1-review')
for detail in expense_raw_verified:
    print('Verified:', detail)
for detail in expense_raw_unavailable:
    print('Not run:', detail)
print('Byte/projection checks do not independently validate manual coding or complete historical coverage.')

Verified: query hash, pagination and both complete saved projections
Verified: original PDF hash
Verified: original: 11 distinct image pairs, dimensions, objects and complete PDF pages
Verified: amendment PDF hash
Verified: amendment: 3 distinct image pairs, dimensions, objects and complete PDF pages
Byte/projection checks do not independently validate manual coding or complete historical coverage.


### Reconcile the archived procurement frame
The committed compact profile comes from a full scan of exactly 56 manifest-selected ZIPs, excluding superseded downloads and subaward members. This cell validates saved provenance, category partitions and totals, not the original ZIP bytes. To repeat the source scan, run `python3 scripts/audit-procurement-bulk-frame.py --scan` with the ignored archives available. A separate reviewed raw excerpt links one blank-type EPA record to a current official IDV identity. It does not classify all blank rows or establish contemporaneous 2024 type history.

In [10]:
bulk_audit = load_script('bulk_audit', 'audit-procurement-bulk-frame.py')
manifest = json.loads(bulk_audit.MANIFEST.read_text())
profile = json.loads(bulk_audit.PROFILE.read_text())
totals = bulk_audit.validate_profile(profile, manifest, bulk_audit.sha256(bulk_audit.MANIFEST))
review = json.loads(bulk_audit.TYPE_REVIEW.read_text())
print('Reviewed blank-type example:', bulk_audit.validate_type_review(review, profile))
assert sum(totals['awardTypeRows'].values()) == totals['rows']
assert sum(totals['obligationSignRows'].values()) == totals['rows']
for key in ('rows', 'childDescriptionRows', 'blankAwardTypeRows', 'missingOffersRows', 'zeroOffersRows', 'afterExclusionCompetitionRows'):
    print(f"{key}: {totals[key]:,} ({100 * totals[key] / totals['rows']:.2f}%)")
print('Signed net obligations, dollars:', totals['netObligationDollars'])
print('Absolute obligations, dollars:', totals['absoluteObligationDollars'])
for partition in profile['strata']:
    source = partition['manifestStratum']
    if int(source['rowCountDrift']):
        print(source['agency'], source['startDate'], source['endDate'], 'count/export difference', source['rowCountDrift'])

Reviewed blank-type example: IDV_B_B
rows: 6,449,101 (100.00%)
childDescriptionRows: 6,205,401 (96.22%)
blankAwardTypeRows: 243,700 (3.78%)
missingOffersRows: 4,505,798 (69.87%)
zeroOffersRows: 351 (0.01%)
afterExclusionCompetitionRows: 665,137 (10.31%)
Signed net obligations, dollars: 704337439585.26
Absolute obligations, dollars: 774172847532.38
Department of Agriculture 2024-07-01 2024-09-30 count/export difference -21
Department of Defense 2024-05-01 2024-05-31 count/export difference -1


### Separate agency-response document pilot
This purposive historical Phase 3 pilot is not part of the Utah corpus. The source JSON preserves original API response hashes, public metadata projections, four PDF hashes, reviewed pages and two source-bound request codings. This offline check verifies saved consistency, not the original PDFs, unread submitted attachments, independent coding accuracy or causal identification. See `docs/comment-response-document-pilot.md` for direct source links, scope and sampling limitations.

In [11]:
response_rows = audit.read('comment-response-document-pilot.csv')
response_sources = json.loads((audit.DATA / 'comment-response-document-source.json').read_text())
response_checks = audit.comment_response_diagnostics(response_rows, response_sources)
assert response_checks['requestRows'] == response_checks['explicitNamedResponses'] == 2
assert response_checks['originalAttachmentsRead'] == response_checks['rateEligibleRequests'] == 0
assert response_checks['independentlyReviewedRequests'] == 0
assert response_checks['causalEffect'] == 'not_identified'
for row in response_rows:
    print(f"{row['commentId']}-A{row['attachmentOrder']}: {row['disposition']}; request PDF {row['requestPdfPage']} / printed {row['requestPrintedPage']}; response PDF {row['responsePdfPage']} / printed {row['responsePrintedPage']}")
print('Request representation: agency-reproduced excerpts; originals unread; independent review pending.')
print('Regulatory text difference observed; individual causal attribution not established.')

EPA-HQ-OAR-2022-0985-1598-A1: clarified_existing_scope; request PDF 458 / printed 440; response PDF 462 / printed 444
EPA-HQ-OAR-2022-0985-1657-A2: disagreed_with_error_claim; request PDF 457 / printed 439; response PDF 462 / printed 444
Request representation: agency-reproduced excerpts; originals unread; independent review pending.
Regulatory text difference observed; individual causal attribution not established.


### Review the complete listed-docket frame without an award join
The thirteen-entry frame comes from the selected decision's File heading, not a complete protest population. Public browser fields were transcribed with native names and solicitation strings; the source JSON preserves the four source-specific base-number aliases and decision footnotes. No original HTML export was available. Fingerprints bind reviewed projections and context, not web-page bytes. Current award descriptions offer two service-area candidates, but all four latest solicitation fields are null and the reopened archived export lacks description/solicitation columns entirely. A decision-reported SAM check is not an independently inspected historical exclusion interval. Independent coding review remains pending.

Inputs: `gao-docket-timing-pilot.csv`, `gao-docket-timing-source.json`, the existing `gao-award-linkage-source.json`, and the committed bulk manifest. Direct public URLs and reviewed locations are in the source JSON and `docs/procurement-source-reconciliation.md`. This offline check neither reopens the public pages nor rescans the original ZIP.

In [12]:
from collections import Counter
from datetime import date
docket_rows = audit.read('gao-docket-timing-pilot.csv')
docket_sources = json.loads((audit.DATA / 'gao-docket-timing-source.json').read_text())
award_sources = json.loads((audit.DATA / 'gao-award-linkage-source.json').read_text())
docket_checks = audit.docket_timing_diagnostics(docket_rows, docket_sources, award_sources, manifest)
assert len(docket_rows) == len({row['docketFileNumber'] for row in docket_rows}) == 13
assert Counter(row['filedDate'] for row in docket_rows) == {'2024-06-24': 6, '2024-06-26': 4, '2024-08-02': 3}
assert all(not row['piid'] for row in docket_rows)
assert all(not row['filedDate'] for row in audit.read('gao-award-linkage-pilot.csv'))
for filed, count in sorted(Counter(row['filedDate'] for row in docket_rows).items()):
    print(f'{filed}: {count} listed docket entries')
for row in docket_rows:
    assert date.fromisoformat(row['filedDate']) <= date.fromisoformat(row['decisionDate']) <= date.fromisoformat(row['postedDate'])
    if ';' in row['solicitationNumbersNative'] or ';' in row['decisionServiceAreas']:
        print(f"{row['docketFileNumber']}: docket solicitations {row['solicitationNumbersNative']}; decision areas {row['decisionServiceAreas']}")
print('One consolidated decision; zero award-specific filing dates or historical exclusion intervals promoted.')
print('13 public-field transcriptions; no archived HTML; independent review pending.')

2024-06-24: 6 listed docket entries
2024-06-26: 4 listed docket entries
2024-08-02: 3 listed docket entries
B-422690.2: docket solicitations 36C10X24R0007;36C10X24R0017; decision areas VISN 22
B-422692.1: docket solicitations 36C10X24R0007;36C10X24R0020; decision areas OGA
B-422693.1: docket solicitations 36C10X24R0007;36C10X24R0014; decision areas VISN 19
B-422693.3: docket solicitations 36C10X24R0014; decision areas VISN 8;VISN 19;VISN 22;OGA
One consolidated decision; zero award-specific filing dates or historical exclusion intervals promoted.
13 public-field transcriptions; no archived HTML; independent review pending.


### Review the complete section without inventing a docket denominator
`comment-section-inventory-source.json` records all seven organization blocks in section 2.5, PDF pages 457-462 (printed 439-444), and a retrospective eleven-request segmentation. Original submissions are not read or version-matched. The previous two-request pilot overlaps this inventory and is not an additional sample. Repeated requests and cross-references are collapsed within an organization; PACCAR's conditional fallback stays separate. The response-page fields for unresolved requests identify the region searched, not an observed answer.

The section was selected after reading its responses. A complete inventory of the agency's selected excerpts cannot establish a docket response rate or causal comparison. Appendix A separately describes 1,011 non-reproduced comments with 2,533 theme incidences; EPA says general responses elsewhere address their topics. These are not 1,011 observed nonresponses. The code below checks counts, mappings and consistency, not independent coding accuracy or source authenticity.

In [13]:
section_rows = audit.read('comment-section-inventory.csv')
section_sources = json.loads((audit.DATA / 'comment-section-inventory-source.json').read_text())
section_checks = audit.comment_section_diagnostics(section_rows, section_sources, response_sources)
assert len(section_rows) == 11
assert len({row['citedCommentId'] for row in section_rows}) == 7
assert sum(bool(row['priorPilotObservationId']) for row in section_rows) == 2
assert all(row['docketRateEligible'] == 'false' for row in section_rows)
for row in section_rows:
    print(f"{row['organizationBlockId']}: {row['requestCode']} -> {row['disposition']} ({row['responseLink']})")
appendix = section_sources['appendixSelectionContext']
unique_comments = sum(row['comments'] for row in appendix['issueCountDistribution'])
theme_incidences = sum(row['issues'] * row['comments'] for row in appendix['issueCountDistribution'])
assert unique_comments == 1011
assert theme_incidences == sum(appendix['themeCounts'].values()) == 2533
assert appendix['nonresponseDenominatorEligible'] is False
print(f'Appendix A: {unique_comments} non-reproduced comments; {theme_incidences} theme incidences, not nonresponses.')
print('Eleven requests include the two prior pilot requests; independent segmentation/coding review remains pending.')

allison: lhd_calculation_error -> disagreed_with_error_claim (explicit_named_response)
carb: require_ci_multipurpose_for_engineless_vehicles -> inconsistent_with_described_resolution (section_resolution_only)
china: explain_table_ii19_method_and_sources -> no_request_specific_disposition (no_request_specific_disposition)
china: explain_international_comparability -> explained_comparability_limit (explicit_named_response)
cummins: tractor_credit_scope -> clarified_existing_scope (explicit_named_response)
dtna: allow_engineering_judgment_subcategory -> consistent_with_collective_revision (named_summary_collective_response)
dtna: remove_zev_averaging_set_limits -> no_request_specific_disposition (no_request_specific_disposition)
paccar: retain_intended_use_subcategory -> consistent_with_collective_revision (named_summary_collective_response)
paccar: conditional_delay_to_my2030 -> conditional_fallback_not_separately_resolved (no_request_specific_disposition)
volvo: reevaluate_vocational_st

### Follow one existing request across sections
`comment-request-followups.json` joins the existing `s25-dtna-averaging` request to section 10.3.2 by the cited submission/attachment and requested action. It is a targeted follow-up, not another sample or a census of that section. The baseline source fingerprint and eleven-request inventory remain unchanged. Seven response-document pages, one proposal page and two final-rule pages were visually reviewed; hashes and PDF/printed page mappings are recorded in the JSON. This offline check validates that saved record, not the original source bytes or an independent reader's interpretation.

The shared response partially aligns with DTNA's request, but the proposal already solicited an interim option. Four facets describe this one request. The published clause's MY2032 use cutoff differs from eligible credit-generation dates, which also include specified pre-2027 advanced-technology credits. No volume cap does not establish a comprehensive no-discount rule. The historical April 2024 clauses do not establish currently operative law. Original-file/version verification, independent coding and causal attribution remain unresolved.

In [14]:
followup_sources = json.loads((audit.DATA / 'comment-request-followups.json').read_text())
followup_checks = audit.comment_followup_diagnostics(followup_sources, section_sources, response_sources)
assert followup_checks['followupReviews'] == followup_checks['crossSectionResponseLinks'] == 1
assert followup_checks['baselineRequests'] == len(section_rows) == 11
assert followup_checks['newIndependentRequests'] == 0
assert followup_checks['requestFacets'] == 4
assert followup_checks['docketRateEligibleRequests'] == followup_checks['independentlyReviewedFollowups'] == 0
assert followup_checks['causalEffect'] == 'not_identified'
followup = followup_sources['reviews'][0]
baseline_request = next(row for row in section_sources['requests'] if row['requestId'] == followup['baselineRequestId'])
assert baseline_request['disposition'] == followup['baselineDisposition'] == 'no_request_specific_disposition'
assert followup['disposition'] == 'partial_alignment_with_interim_flexibility'
assert followup['proposalAlreadySolicitedOption'] and not followup['proposalWasAdoptedRule']
assert followup['specifiedPre2027AdvancedTechnologyCreditsIncluded']
assert followup['historicalFinalUseThroughModelYear'] == 2032
assert sum(len(document['reviewedPages']) for document in followup_sources['documents'].values()) == 10
print(f"{followup['baselineRequestId']}: section 2.5 {baseline_request['disposition']} -> section 10.3.2 {followup['disposition']}")
for facet in followup['facets']:
    print(f"{facet['facet']}: {facet['assessment']}; final PDF pages {facet['evidencePdfPages']}")
print('One existing request; four facets; ten reviewed source pages; no new independent observations.')
print('Proposal already solicited the interim option; no individual causal effect identified.')

s25-dtna-averaging: section 2.5 no_request_specific_disposition -> section 10.3.2 partial_alignment_with_interim_flexibility
vehicle_transfer_permission: aligned_with_limited_vehicle_transfer; final PDF pages [337, 351]
program_life_duration: narrower_time_limit; final PDF pages [337]
traded_credit_eligibility: aligned_subject_to_retained_restrictions; final PDF pages [337, 351]
no_credit_discount: no_separate_discount_disposition_verified; final PDF pages [337, 351]
One existing request; four facets; ten reviewed source pages; no new independent observations.
Proposal already solicited the interim option; no individual causal effect identified.


### Recover original-action fields without overwriting the docket baseline

The September 13 UTC follow-up retains all four awards from the existing one-decision pilot. Each public history response has complete pagination; four contract-specific ZIPs passed CRC checks. The 26 downloaded transaction IDs agree with the API histories. The original CSV rows retain descriptions, full award/action keys, reported offers, solicitation fields and source timestamps. Requests, source URLs, hashes and original row locations are in `gao-original-action-review.json`. A failed general-download job is retained as an acquisition failure, not a zero-row outcome.

Two original descriptions explicitly name VISN 19 and VISN 22. Two other original potential-total values uniquely match the decision's OGA and VISN 8 amounts after its stated whole-dollar rounding, supplying amount-only candidates rather than verified solicitation identities. The provisional crosswalk has sixteen pairs across thirteen dockets; one supplemental case spans all four areas. The explicit VISN 22 record rounds to $100 below GAO's order value, as its footnote 15 notes. Do not correct the source amount or use a general fuzzy-match tolerance. All four native solicitation identifiers remain blank. Generic references to other government agencies and residual order are not mapping evidence. The candidate VISN 8 record reports four offers while GAO reports five timely proposals; the other three counts agree, but no underlying bids or source-definition equivalence are independently verified. The old timing ledgers remain unchanged; these are newly retrieved records about 2024 actions, not a contemporaneous archive, a representative SAM panel, independently reviewed timing, historical exclusion evidence or a causal estimate.

The next cell checks the saved projections and derives the provisional crosswalk. It does not authenticate raw source bytes or independent coding. When the ignored acquisition files are present, the optional raw check additionally reopens the ZIPs and compares every preserved original field and full original-row fingerprint. The notebook remains runnable without those local source files; absence is reported rather than treated as raw verification.


In [15]:
original_review = json.loads((audit.DATA / 'gao-original-action-review.json').read_text())
award_baseline = json.loads((audit.DATA / 'gao-award-linkage-source.json').read_text())
docket_baseline = json.loads((audit.DATA / 'gao-docket-timing-source.json').read_text())
original_summary, provisional_pairs = audit.original_action_diagnostics(
    original_review, award_baseline, docket_baseline, audit.read('gao-docket-timing-pilot.csv'))
assert original_summary['reviewedAwards'] == 4
assert original_summary['historyRows'] == 26
assert original_summary['provisionalAwardDocketPairs'] == 16
assert original_summary['distinctLinkedDockets'] == 13
assert original_summary['awardSpecificDatesPromoted'] == 0
assert original_summary['reportedOffersMinusTimelyProposals'] == {'36C10X24N0089': -1}
print(json.dumps(original_summary, indent=2))
for pair in provisional_pairs:
    print(pair['piid'], pair['serviceArea'], pair['decisionCaseId'], pair['filedDate'], pair['mappingBasis'], pair['status'])

{
  "reviewedAwards": 4,
  "historyRows": 26,
  "originalActions": 4,
  "blankSolicitationIdentifiers": 4,
  "explicitOriginalServiceAreas": 2,
  "amountBasedServiceAreaCandidates": {
    "36C10X24N0088": "OGA",
    "36C10X24N0089": "VISN 8"
  },
  "roundedPotentialValueMinusDecisionDollars": {
    "36C10X24N0074": 0,
    "36C10X24N0088": 0,
    "36C10X24N0089": 0,
    "36C10X24N0108": -100
  },
  "reportedOffersMinusTimelyProposals": {
    "36C10X24N0089": -1
  },
  "independentlyReviewedMappings": 0,
  "provisionalAwardDocketPairs": 16,
  "distinctLinkedDockets": 13,
  "originalReportedOffers": {
    "36C10X24N0074": "4",
    "36C10X24N0088": "2",
    "36C10X24N0089": "4",
    "36C10X24N0108": "4"
  },
  "awardSpecificDatesPromoted": 0,
  "historicalExclusionsPromoted": 0,
  "causalEffect": "not_identified"
}
36C10X24N0074 VISN 19 B-422693 2024-06-24 original_description_bridge_pending_independent_review provisional_not_promoted
36C10X24N0074 VISN 19 B-422693.2 2024-06-26 original_de

In [16]:
import hashlib
import io
import zipfile

raw_directory = root / 'data/raw/gao-original-action-review'
raw_verified = []
raw_unavailable = []
offer_context = {}
for award in original_review['awards']:
    suffix = award['piid'][-4:]
    archive_path = raw_directory / f'gao-{suffix}.zip'
    history_path = raw_directory / f'gao-{suffix}-history.json'
    if not archive_path.exists() or not history_path.exists():
        raw_unavailable.append(award['piid'])
        continue
    assert hashlib.sha256(archive_path.read_bytes()).hexdigest() == award['archiveSha256']
    history_bytes = history_path.read_bytes()
    assert hashlib.sha256(history_bytes).hexdigest() == award['historyRawSha256']
    assert json.loads(history_bytes) == award['historyResponse']
    with zipfile.ZipFile(archive_path) as archive:
        assert archive.testzip() is None
        member_bytes = archive.read(award['transactionMember'])
    assert hashlib.sha256(member_bytes).hexdigest() == award['transactionMemberSha256']
    records = list(csv.DictReader(io.StringIO(member_bytes.decode('utf-8-sig'))))
    assert len(records) == award['transactionRowCount']
    assert len(records[0]) == award['transactionColumnCount']
    assert [r['contract_transaction_unique_key'] for r in records] == award['transactionIds']
    original = records[award['originalCsvRowIncludingHeader'] - 2]
    assert audit.source_fingerprint(original) == award['originalFullRowFingerprint']
    assert {k: original[k] for k in award['originalProjection']} == award['originalProjection']
    offer_headers = sorted(k for k in original if 'offer' in k.lower())
    assert offer_headers == ['number_of_offers_received']
    offer_context[award['piid']] = {'offerNamedHeaders': offer_headers, **{
        k: original[k] for k in ('fair_opportunity_limited_sources_code',
                                'idv_type_code', 'multiple_or_single_award_idv_code')}}
    raw_verified.append(award['piid'])
print('Raw ZIP/member/row and history checks:', len(raw_verified))
print('Raw source checks unavailable for:', raw_unavailable)
print('Offer-field context from verified raw original rows:')
for piid, context in offer_context.items():
    print(piid, context)
print('Offer-source provenance and historical dictionary rules remain unresolved; child-row IDV blanks do not classify the parent.')
print('Independent source/coding review remains pending.')

Raw ZIP/member/row and history checks: 4
Raw source checks unavailable for: []
Offer-field context from verified raw original rows:
36C10X24N0074 {'offerNamedHeaders': ['number_of_offers_received'], 'fair_opportunity_limited_sources_code': 'FAIR', 'idv_type_code': '', 'multiple_or_single_award_idv_code': ''}
36C10X24N0088 {'offerNamedHeaders': ['number_of_offers_received'], 'fair_opportunity_limited_sources_code': 'FAIR', 'idv_type_code': '', 'multiple_or_single_award_idv_code': ''}
36C10X24N0089 {'offerNamedHeaders': ['number_of_offers_received'], 'fair_opportunity_limited_sources_code': 'FAIR', 'idv_type_code': '', 'multiple_or_single_award_idv_code': ''}
36C10X24N0108 {'offerNamedHeaders': ['number_of_offers_received'], 'fair_opportunity_limited_sources_code': 'FAIR', 'idv_type_code': '', 'multiple_or_single_award_idv_code': ''}
Offer-source provenance and historical dictionary rules remain unresolved; child-row IDV blanks do not classify the parent.
Independent source/coding review

## Takeaways
The source snapshot contains 1,119 LDA issue rows across 427 filings and six API-attributed candidates, each with nine native pre-enactment reporting periods under the unvalidated source attribution. All three flagged posting-date anomalies now have embedded-cover reviews, but no corrected API posting times or complete version families are established. Fifteen restored AdvaMed filings include a 2008Q2 amendment that still needs version-family review. The metadata contains 267 income reports, 89 expense reports and 71 filings with neither amount; 63 expense reports lack API accounting methods. Seventy-four actor-period cells contain both reporting sides, which must not be added into organizational totals. The two reviewed 2003 covers use Methods C and A respectively, so being in the same LDA source does not establish a common federal-only outcome.

The four-PAC acquisition frame retains 48 expected half-years: GASPAC, AdvaMed and AAJ each supply twelve, while all twelve Benefits Council periods remain unresolved. AAJ's final half-year depends on the documented supplemental-filing review; omitting that decision yields 35 usable half-years. There are 27 pre-event, three excluded event-straddling and six post-event observations pooled across three PACs. Available outcomes and unassigned reform exposure cannot identify a treatment-control substitution effect. The legacy candidate-contributions field includes Form 3X contributions to other political committees, not solely candidates. The two comment pilot observations have no adjudicated individual-comment links. The 2026 GAO worklist remains unlinked. Separately, four FY2024 award links from one consolidated decision match archived bulk and current original-action records, but none appears in the 28,103-row small panel. That panel contains 134 repeated partial keys, not proven duplicate actions. The pilot supports source reconciliation, not protest rates or a capture effect.

The bulk profile preserves 6,205,401 explicit child-action descriptions and 243,700 unresolved blank types, including one verified vehicle identity. Missing offers cover 69.87% of exported rows and are distinct from 351 explicit source zeros. The 665,137 after-exclusion competition rows describe a competition procedure, not vendor debarment. Agriculture Q4 and Defense May account for the 22-row count/export discrepancy, whose cause and record identities remain unknown. Matching archived hashes and agency/date coverage does not certify a complete, unique-action SAM comparison frame.

The separate Phase 3 pilot documents two named responses from one rule: Cummins receives a clarification of existing tractor-credit scope, while EPA disagrees with Allison's alleged NPRM error. The response document reproduces the requests; the original submitted files remain unread. The final Std definition differs from the proposal, but the collective method change and later publication correction cannot be attributed to either request. No response rate or causal estimate is computed from these selected cases.

The unchanged docket ledger contains all thirteen entries listed in the FY2024 consolidated decision, with six June 24, four June 26 and three August 2 filing dates. Three entries list multiple solicitations; one spans all four service areas. Its original limited-export evidence is supplemented by the separate original-action follow-up below, but baseline award-specific timing remains unassigned. The decision's reported SAM review has no verified check date or inspected underlying records, so it does not establish exclusion intervals. Docket deadlines and displayed posting dates remain distinct from the decision date.

The full-section inventory exposes distinctions hidden by organization-level acceptance labels: CARB requests a Multi-Purpose restriction inconsistent with the described resolution; DTNA's averaging-set exemption has no separate disposition in this section; PACCAR's MY2030 fallback is conditional on a proposal EPA says it did not finalize. Four requests align with a shared revision, not four independently caused policy changes. The appendix's theme totals use a different unit from its comment count and do not define nonresponses.

The later section-10.3.2 review resolves a documentary gap for the existing DTNA averaging-set request, without revising the original section-2.5-only coding. Partial alignment includes time-limited vehicle-credit transfers and trading subject to retained restrictions, not a program-life exemption. No comprehensive no-discount finding is verified. EPA had already invited comment on an interim option in the proposal, so the final choice cannot be attributed to DTNA from this sequence. The ten reviewed pages span three historical documents; they are neither original-file verification nor independent review.

The three complete cover images recover the receipt dates 2003-07-31, 2004-02-24 and 2005-01-19. Each full Senate ID matches the API registrant/client-relationship components despite two registrant-name differences. APGA's candidate cover reports $100,000 using Method A but has unchecked threshold boxes; ATLA's 2003H2 cover reports $2,380,000 using Method A. These methods apply only to the reviewed filings. Shea & Gardner's selected income category is below $10,000 without a point amount, and its January 5, 2005 termination date is absent from the API field. The raw API zero and timestamps remain unchanged; the other 24 source-zero filings and 424 unflagged dates are not cleared by this selected review.

The LDA family queue covers 22 multi-record candidate groups containing 46 filings, with eight different-amount flags, three posting ties and two latest-posting missing-amount flags. A selected APGA amendment contains only one distinct issue image, not a financial cover. Both energy-page scans name FERC but their API contact lists do not reproduce that agency list. That APGA review selects no final version, spending amount or exposure classification.

A follow-up reviews all 21 frozen NVG filings: seventeen full embedded PDF covers and four electronic HTML identity/timing sections. Six name America Votes, including one registration, and are ineligible for AAJ attribution. Five mixed-client candidate groups are not mergeable. Two cover years disagree with API years; Senate suffixes cannot be globally repaired. The earlier America Votes House-ID transcription is corrected to 35960015 against the unchanged source image. The other fifteen records are not cleared as final actor-period outcomes. Raw metadata remains unchanged, and independent review is pending.

This notebook verifies reproducible profiling and transformation, not historical coverage, causal identification, independent manual coding, or representative procurement sampling. It does not alter the Java model or promote calibration claims.

The original-action follow-up supplies two explicit service-area bridges and two rounded-value candidates: sixteen provisional pairs, not complete award timing or independent events. The candidate VISN 8 offer/proposal count discrepancy, four blank solicitation identifiers, historical-vintage limitations and pending independent review remain explicit.

A separate AAJ 2005H1 pair supports one reported $4,020,000 Method A expense through matching covers and an explicit unchanged-expense letter. All fourteen distinct downloaded images were reviewed. Empty amendment API activities and two original contact discrepancies remain unresolved. This field-specific adjudication does not select a whole final filing or clear actor totals, controls or independent review.

The Benefits Council paper review matches all 24 frozen reports to a separate 280-record filing inventory and recovers 21 current plus three superseded version flags. Seven selected report PDFs reveal three native-form discrepancies and three downloaded packets without a disbursement-summary page. Version recovery adds no complete half-year: eleven periods fail on missing outcomes and 2006H1 first fails on its source-confirmed date gap. Raw dates/amounts and the 36 usable half-years remain unchanged; independent review remains pending.